### ISSUE TITLE
F-TEID unmarshalling and Enterprise IEs

### ISSUE THREAD

**Issue by vulcanfan** [wmnsk/go-pfcp#73]
Hi,

Many thanks for the last fixes and merge.

Looking further, I have a couple more observations:

1/  In f-teid.go at line 215 (v0.0.8) where the CH bit is set, I don't believe there is meant to be an IP address (v4 or v6) expected.  If the IPv4 or IPv6 bit is set then that means that the UPF should allocate/choose an IP address, not that one is provided in this message.  

2/  In ie.go UnmarshalBinary(), I believe:
```
    if i.IsVendorSpecific() && l >= 6 {
        i.EnterpriseID = binary.BigEndian.Uint16(b[4:6])
        offset += 2
    }
```

should be 
```
    if i.IsVendorSpecific() && l >= 6 {
        i.EnterpriseID = binary.BigEndian.Uint16(b[4:6])
        offset += 2
        i.Length -= 2   
    }
```
Otherwise it panics lower down because the slice upper range is out of bounds.

Many thanks

**Comment by wmnsk** [comment]
Sorry leaving this for so long! Please give me the pointer to why you think it's wrong makes it pretty much easier to investigate the issue, hopefully.

1:
Yup, the IP addresses are expected to be absent in that case, but it also checks the V4/V6 flags, which would be invalid if set. Do we have any real problem for that?

2:
The spec says the length excludes the first four octets, so I think we don't need to minus two.

```
Length: this field contains the length of the IE excluding the first four octets, which are common for all IEs 
(Type and Length) and is denoted "n" in Figure 8.1.1-1 and in Figure 8.1.1-2. Bit 8 of the lowest numbered octet 
is the most significant bit and bit 1 of the highest numbered octet is the least significant bit.
```

**Comment by vulcanfan** [comment]
Hi,

Sure, no problem.  To respond to your points:

1/  The spec for octet 5 says: 
"If this bit is set to "1", then the TEID, IPv4 address and IPv6 address fields shall not be present and the UP function shall assign an F-TEID with an IP4 or an IPv6 address if the V4 or V6 bit is set respectively."

So it is expected that the IPv4 bit is set yet there be no actual IP address is there.  The bit in this case is not there to denote the presence of the IP address in subsequent bytes, but that the UPF needs to allocate an IPv4 address itself.  The code currently increments offset by 4 bytes if CH is set and IPv4 is set.

As currently written, for example, 'f.ChooseID = b[offset:]' (line 235) is executed even if HasChId() is not set (but HasCh() is set instead, because of the || in the if statement).  The offset will calculated wrong (because of the assumption above of there being an IP address present) then there is a likely slice index OOR panic.

2/  The payload length as recorded in the IE does indeed omit the first 4 octets, in the code at line 334 'i.Length' is indeed initially set to the the payload length.  However in line 346 i.Length is semantically used as the 'length remaining' after whatever 'offset' is set to. 

So I think the issue is really in line 346, being 'i.Payload = b[offset : offset+int(i.Length)]'.  Increasing offset by 2 without reducing i.Length by 2 will result in the end index of that slice being out of bounds, ie beyond the end of the slice, so panics.  In fact that slice assignment only works if offset is equal to 4 (since i.Length happens to be effectively initialised as len(b) - 4 already as that's what the encoded length hopefully is).  Incrementing offset beyond 4, without reducing what i.Length holds, makes that i.Payload = assignment panic.

Thanks

**Comment by wmnsk** [comment]
Fixed finally and added a new tag. Thank you for your report and patience :)

**Comment by wmnsk** [comment]
Ah, no I haven’t addressed the second point(enterprise id). Let me take a look later.

**Comment by wmnsk** [comment]
Fixed in #80.

**Comment by wmnsk** [comment]
Let me know if you need anything additionally.
